### PR TITLE
BUG: fix reading of compressed nrrd

### PR DESCRIPTION
**Pull request by Punzo**
when reading compressed files, the new reading API (streaming) has not to be used (very poor performances).

### CODE REVIEW

File: Sequences/MRML/vtkMRMLVolumeSequenceStorageNode.cxx [hunk 1]
```diff
@@ -164,6 +164,21 @@ int vtkMRMLVolumeSequenceStorageNode::ReadDataInternal(vtkMRMLNode* refNode)
   // Read and copy the data to sequence of volume nodes
 #ifdef NRRD_CHUNK_IO_AVAILABLE
   int numberOfFrames = reader->GetNumberOfImages();
+  vtkImageData* imageData = NULL;
+  vtkNew<vtkImageExtractComponents> extractComponents;
+  if (reader->GetIsCompression())
```
Comment (lassoan) [inline Sequences/MRML/vtkMRMLVolumeSequenceStorageNode.cxx]: Property name can be either IsCompressed or UseCompression (not IsCompression). UseCompression is typically used when we can choose to enable/disable compression (e.g., in file writers), so maybe IsCompressed is the better name.

**Comment by Punzo** [comment]
@lassoan

**Review by lassoan — COMMENTED** [review]
We have already set reader->ReadImageListAsMultipleImagesOn(); which means that images must be returned as separate volumes and not as a single multi-component image. Probably the simplest and cleanest solution is to to move extract components filter into the nrrd reader.

File: Sequences/MRML/vtkMRMLVolumeSequenceStorageNode.cxx (hunk 1 above)
Comment (Punzo) [reply Sequences/MRML/vtkMRMLVolumeSequenceStorageNode.cxx]: ok, fixed already

**Comment by Punzo** [comment]
> We have already set reader->ReadImageListAsMultipleImagesOn(); which means that images must be returned as separate volumes and not as a single multi-component image. Probably the simplest and cleanest solution is to to move extract components filter into the nrrd reader.

ok, I'll see, if I can move everything in the nrrd reader

**Comment by Punzo** [comment]
>>We have already set reader->ReadImageListAsMultipleImagesOn(); which means that images must be returned as separate volumes and not as a single multi-component image. Probably the simplest and cleanest solution is to to move extract components filter into the nrrd reader.

>ok, I'll see, if I can move everything in the nrrd reader

I could not move the extract component filter into the reader for various reasons:
1) if you want to read only one volume at time (i.e. GetOutput return only one vtkImageData) you have to read the full file everytime (and byte skipping is very slow for compressed datasets, therefore this should be avoided)
2) if you return all the volumes at a single GetOutput call, there is the issue that we have to return a collection of vtkImageData in some way (not clear how, and anyway we have to change the code in Sequence for handling this special case) (I mean: is it possible to overload the GetOutput method of a reader? in the reader we call ExecuteDataWithInformation, so I think we will have to modify a lot of things) 

Therefore, I propose that we do not explicit check if the file is compressed in Sequence, this will be done internally in the reader, but still leave the extract component filter in the Sequence method. When Sequence::ReadDataInternal calls ReadImageListAsMultipleImagesOn, this will return a bool if reading the data as a single multi-component image is allowed (i.e.  non compressed). Please see the implementation in https://github.com/SlicerRt/Sequences/pull/59/files#diff-12045c4e7361ade3dec29f0e249fd75cR95 and https://github.com/ProxSim/Slicer/pull/10/files#diff-354b8027cad5352096468535a16667b8R239

**Comment by lassoan** [comment]
Indeed, gzseek is slow and there is no simple solution. For future reference, here are a few interesting links on the topic:
- https://stackoverflow.com/questions/429987/compression-formats-with-good-support-for-random-access-within-archives
- https://github.com/madler/zlib/blob/master/examples/zran.c
- http://blastedbio.blogspot.com/2011/11/bgzf-blocked-bigger-better-gzip.html
- http://blastedbio.blogspot.com/2011/11/random-access-to-bzip2.html
- http://www.bzip.org/1.0.3/html/memory-management.html

We could process the whole compressed data block by block. After a block is uncompressed,, we would copy voxels into correct location in individual volumes.

Return multiple volumes: The nicest would be to provide filter output as temporally changing data. Please have a look at TestTemporalSupport.cxx VTK example. What do you think?

The advantage of using the temporal data API would be that in the long term it could allow us to read data from file on demand, without the need to read all frames into memory in advance. Startup would be fast and if there are memory concerns then for uncompressed files we could unload the image data from memory when not used. It would work by instantiating a single image source (nrrd reader) and then for each volume node: we would create a time filter, set image source as input, set the time point to be used, and connect the time filter's output to the input connection of the volume node.

The current solution (multi-component image or not, depending on compression) could be probably acceptable for CPA, but would be too messy to be integrated into Slicer core. So, if you think it would be too much time to use the temporal API then we can do it in two steps - integrate the current version, and then switch to temporal API later, when we integrate this into Slicer core.

**Comment by Punzo** [comment]
Thanks for the feedback Andras.

I was also looking for better random access utilities for compressed data. These would solve everything because we can use again always the new teem reading/writing streaming API (both for compressed and uncompressed data). Although to have better random access for compressed data, we have most likely to rewrite completely the compression utilities in teem: https://github.com/Slicer/teem/blob/e4746083c0e1dc0c137124c41eca5d23adf73bfa/src/nrrd/gzio.c

>We could process the whole compressed data block by block. After a block is uncompressed,, we would copy voxels into correct location in individual volumes.

yes definitively, the vkt temporal filter (https://github.com/Kitware/VTK/blob/master/Common/ExecutionModel/Testing/Cxx/TestTemporalSupport.cxx) will be a more elegant solution respect to the ReadImageListAsMultipleImages interface that I have implemented in the nrrdReader of Slicer.

However, this will not fix the issue for the compressed file. In fact, to be used, the temporal filter needs that the data can be streamed from the source. For uncompressed data we can do it already in teem (so the temporal filter can replace my interface in the nrrdReader), but for uncompressed data, we have to find a solution in teem to stream the data with decent performances.

Essentially what we need, it is to replace these lines:
https://github.com/Slicer/teem/blob/slicer-2015-04-14-r6245/src/nrrd/encodingGzip.c#L147-L155

with random access utilities for compressed files.

Then if we want to use also temporal filter for the Slicer interface will be good, but I think it is another matter.

For CPA depends John if he wants to allocate time for this I guess. I mean if implementing:
1) random access for compressed files to allow streaming with good performances in TEEM
2) temporal filter as new interface to read/write data in the nrrdReader in SLICER

are necessary to have the new API integrated in Slicer, we will need to do it anyway at a certain point  for updating the Slicer branch of CPA. However I think that this will takes few days and I don't know if John has higher priority tasks at the moment.

**Comment by jdx-john** [comment]
Since CPA doesn't create compressed nrrd files or need to sort them, I'm wary to commit ProxSim resources to improving slicer... I'm not sure if we already have a solution that teem will work with gzip need or not? From my perspective as long as we don't break teem/slicer that's our only obligation.

I'm a bit lost as I thought we already figured out a fix to the bug we found

**Comment by Punzo** [comment]
> I'm a bit lost as I thought we already figured out a fix to the bug we found

the solution (this PR and https://github.com/ProxSim/Slicer/pull/10) was to not use the streaming for compressed data. However, the solution can't be "clean" and involves checks if the data are compressed and some portions of code under if. This in general should be avoided, but it was the only solution without fixing the streaming performance of compressed files (which requires better handling at Teem level).

**Comment by lassoan** [comment]
> Since CPA doesn't create compressed nrrd files

Using the proposed temporal data API would have multiple advantages for CPA (regardless of compressed/non-compressed storage):

- It would allow setting up the whole data pipeline (from reading from disk to volume rendering). The data could be pulled on request (or could be pulled by background threads in the reader), so you would not need to wait for the complete data set to be read into memory. We could set up things so that on low-memory configurations we would only keep one frame in memory.
- It would be a clean solution that can be merged into Slicer core, so that we don't have to maintain it separately when anything in Slicer, VTK, or teem library changes.

**Comment by Punzo** [comment]
> It would allow setting up the whole data pipeline (from reading from disk to volume rendering). The data could be pulled on request (or could be pulled by background threads in the reader), so you would not need to wait for the complete data set to be read into memory. We could set up things so that on low-memory configurations we would only keep one frame in memory.

that's true, I didn't think about it

**Comment by lassoan** [comment]
> However, this will not fix the issue for the compressed file. In fact, to be used, the temporal filter needs that the data can be streamed from the source. 

It would solve the issue of compressed file, if we cached all image data for all time points in the reader. When you connect a display pipeline to the reader, the image data is not deep-copied. If the file is not compressed (and so we have relatively fast random access) we could even decide to not cache all data in memory but read from disk when rendering pipeline requests it.

**Comment by Punzo** [comment]
P.S.: just to be clear: I think too that implementing the Andras' suggestions will be good for CPA (especially for updating the Slicer branch of CPA).

**Comment by lassoan** [comment]
Merging it now. We can decide about final design after testing everything for a few weeks.

**Comment by jcfr** [comment]
Should some of these notes be transfered on the Slicer wiki in the [Labs](https://www.slicer.org/wiki/Documentation/Labs) page ? Or on this project wiki .. ?

**Comment by lassoan** [comment]
Good idea, we can create a lab page. For this to work we need changes in vtkNRRDTeemReader in Slicer core and also the Teem library.

**Comment by Punzo** [comment]
wrote a very small Lab page regarding this https://www.slicer.org/wiki/Documentation/Labs/NrrdReading_Writing_Optimizations

I have to put aside for the moment further dev on this, but I think we will go on later on. Anyone is welcome to update it, or to continue the dev.

**Pull request merged**